micrometer-tracing/src/main/java/io/micrometer/tracing/contextpropagation/ObservationAwareBaggageThreadLocalAccessor.java

```diff
@@ -0,0 +1,295 @@
+/*
+ * Copyright 2024 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.tracing.contextpropagation;
+
+import io.micrometer.common.lang.NonNull;
+import io.micrometer.common.util.internal.logging.InternalLogger;
+import io.micrometer.common.util.internal.logging.InternalLoggerFactory;
+import io.micrometer.context.ThreadLocalAccessor;
+import io.micrometer.observation.Observation;
+import io.micrometer.observation.ObservationRegistry;
+import io.micrometer.tracing.Baggage;
+import io.micrometer.tracing.BaggageInScope;
+import io.micrometer.tracing.Span;
+import io.micrometer.tracing.Tracer;
+import io.micrometer.tracing.Tracer.SpanInScope;
+import io.micrometer.tracing.handler.TracingObservationHandler;
+
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
+/**
+ * {@link ThreadLocalAccessor} used to propagate baggage via {@link BaggageToPropagate}.
+ *
+ * @author Marcin Grzejszczak
+ * @since 1.2.2
+ */
+public class ObservationAwareBaggageThreadLocalAccessor implements ThreadLocalAccessor<BaggageToPropagate> {
+
+    private static final InternalLogger log = InternalLoggerFactory
+        .getInstance(ObservationAwareBaggageThreadLocalAccessor.class);
+
+    final Map<Thread, BaggageAndScope> baggageInScope = new ConcurrentHashMap<>();
+
+    /**
+     * Key under which Micrometer Tracing Baggage accessor is being registered.
+     */
+    public static final String KEY = "micrometer.tracing.baggage";
```

Comment:
I'm thinking if we should use a String here or a class instead as the key.

Reply:
For observation we do `micrometer.observation`, for Spans we're doing `micrometer.tracing` so I thought that for Baggage we could do `micrometer.tracing.baggage`